atxsdata/data.go

```diff
@@ -0,0 +1,180 @@
+package atxsdata
+
+import (
+	"sync"
+
+	"go.uber.org/atomic"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+)
+
+type ATX struct {
+	Node               types.NodeID
+	Weight             uint64
+	BaseHeight, Height uint64
+	Nonce              types.VRFPostIndex
+	Malicious          bool
+}
+
+type Opt func(*Data)
+
+// WithCapacity sets the number of epochs from the latest applied
+// that cache will maintain in memory.
+func WithCapacity(capacity types.EpochID) Opt {
+	return func(cache *Data) {
+		cache.capacity = capacity
+	}
+}
+
+// WithCapacityFromLayers sets capacity to include all layers in the window.
+func WithCapacityFromLayers(window, epochSize uint32) Opt {
+	capacity := window / epochSize
+	if window%epochSize != 0 {
+		capacity++
+	}
+	return WithCapacity(types.EpochID(capacity))
+}
+
+func New(opts ...Opt) *Data {
+	cache := &Data{
+		capacity:  2,
+		malicious: map[types.NodeID]struct{}{},
+		epochs:    map[types.EpochID]epochCache{},
+	}
+	for _, opt := range opts {
+		opt(cache)
+	}
+	return cache
+}
+
+type Data struct {
+	evicted atomic.Uint32
+
+	// number of epochs to keep
+	// capacity is not enforced by the cache itself
+	capacity types.EpochID
+
+	mu        sync.RWMutex
+	malicious map[types.NodeID]struct{}
+	epochs    map[types.EpochID]epochCache
+}
+
+type epochCache struct {
+	index map[types.ATXID]*ATX
+}
+
+func (d *Data) Evicted() types.EpochID {
+	return types.EpochID(d.evicted.Load())
+}
+
+func (d *Data) IsEvicted(epoch types.EpochID) bool {
+	return d.evicted.Load() >= epoch.Uint32()
+}
+
+// OnEpoch is a notification for cache to evict epochs that are not useful
+// to keep in memory.
+func (d *Data) OnEpoch(applied types.EpochID) {
+	if applied < d.capacity {
+		return
+	}
+	evict := applied - d.capacity
+	if d.IsEvicted(evict) {
+		return
+	}
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if d.evicted.Load() < evict.Uint32() {
+		d.evicted.Store(evict.Uint32())
+	}
+	for epoch := range d.epochs {
+		if epoch <= evict {
+			delete(d.epochs, epoch)
+			atxsCounter.DeleteLabelValues(epoch.String())
+		}
+	}
+}
+
+func (d *Data) Add(
+	epoch types.EpochID,
+	node types.NodeID,
+	atx types.ATXID,
+	weight, baseHeight, height uint64,
+	nonce types.VRFPostIndex,
+	malicious bool,
+) {
+	if d.IsEvicted(epoch) {
+		return
+	}
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	ecache, exists := d.epochs[epoch]
+	if !exists {
+		ecache = epochCache{
+			index: map[types.ATXID]*ATX{},
+		}
+		d.epochs[epoch] = ecache
+	}
```

Comment:
It could re-add an evicted epoch if `d.OnEpoch(epoch+d.capacity)` is called between the `d.IsEvicted(epoch)` check and the lock. Why not bring the check under the lock?
The re-added epoch would probably stay in the `d.epochs` map forever.

```suggestion
	d.mu.Lock()
	defer d.mu.Unlock()
	if d.IsEvicted(epoch) {
		return
	}
	ecache, exists := d.epochs[epoch]
	if !exists {
		ecache = epochCache{
			index: map[types.ATXID]*ATX{},
		}
		d.epochs[epoch] = ecache
	}
```

Reply:
i wanted to be "smart" and make OnEpoch non-blocking (basically noop if it has nothing to evict) so that it doesn't interfere with read locks, but it is generally rare and won't make substantial difference.

it won't store it forever but for one more epoch, because of this

```go
for epoch := range d.epochs {
		if epoch <= evict {
			delete(d.epochs, epoch)
			atxsCounter.DeleteLabelValues(epoch.String())
		}
	}
```

Reply:
btw i moved everything under the lock